Iss53squash

hustf opened the pull request:
Note that no docbuild was attempted. "Exteral documentation" has a few names replaced.

See commit message https://github.com/kwehmeyer/Spotify.jl/commit/a47210660e523483c4e711e8abfbe3f734fa0d83

hustf committed:
Preparing to delete unmaintainable generalized calls
modified:   Project.toml
renamed:    example/csv_of_saved_tracks.jl -> example/csv_of_liked_tracks.jl
modified:   example/list_my_tracks_beat_per_minute.jl
new file:   example/script_create_playlists_from_referred_tracks_by_bpm.jl
new file:   example/script_list_liked_tracks.jl
deleted:    example/script_list_my_tracks.jl
modified:   src/Spotify.jl
new file:   src/authorization/access_local_credentials.jl
modified:   src/authorization/client_credentials.jl
modified:   src/authorization/implicit_grant_flow.jl
new file:   src/by_console_doc/README.md
renamed:    src/by_reference_doc/browse.jl -> src/by_console_doc/browse.jl
renamed:    src/by_reference_doc/follow.jl -> src/by_console_doc/follow.jl
renamed:    src/by_reference_doc/library.jl -> src/by_console_doc/library.jl
renamed:    src/by_reference_doc/personalization.jl -> src/by_console_doc/personalization.jl
modified:   src/by_reference_doc/README.md
modified:   src/by_reference_doc/albums.jl
renamed:    src/by_reference_doc/artist.jl -> src/by_reference_doc/artists.jl
new file:   src/by_reference_doc/categories.jl
modified:   src/by_reference_doc/episodes.jl
modified:   src/by_reference_doc/player.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/search.jl
modified:   src/by_reference_doc/shows.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/by_reference_doc/users.jl
modified:   src/export_structure.jl
new file:   src/lookup_containers/all_scopes_vector.jl
new file:   src/lookup_containers/response_codes_dic.jl
new file:   src/request.jl
modified:   src/types.jl
deleted:    src/util/request.jl
modified:   src/util/utilities.jl
modified:   test/Project.toml
new file:   test/function_default_arguments_dic.jl
modified:   test/generalize_calls.jl
modified:   test/interactive_default_calls.jl
modified:   test/runtests.jl
modified:   test/test_browse.jl
modified:   test/test_int_format_strings.jl
modified:   test/test_users.jl

hustf committed:
Preparing to delete unmaintainable generalized calls
modified:   Project.toml
renamed:    example/csv_of_saved_tracks.jl -> example/csv_of_liked_tracks.jl
modified:   example/list_my_tracks_beat_per_minute.jl
new file:   example/script_create_playlists_from_referred_tracks_by_bpm.jl
new file:   example/script_list_liked_tracks.jl
deleted:    example/script_list_my_tracks.jl
modified:   src/Spotify.jl
new file:   src/authorization/access_local_credentials.jl
modified:   src/authorization/client_credentials.jl
modified:   src/authorization/implicit_grant_flow.jl
new file:   src/by_console_doc/README.md
renamed:    src/by_reference_doc/browse.jl -> src/by_console_doc/browse.jl
renamed:    src/by_reference_doc/follow.jl -> src/by_console_doc/follow.jl
renamed:    src/by_reference_doc/library.jl -> src/by_console_doc/library.jl
renamed:    src/by_reference_doc/personalization.jl -> src/by_console_doc/personalization.jl
modified:   src/by_reference_doc/README.md
modified:   src/by_reference_doc/albums.jl
renamed:    src/by_reference_doc/artist.jl -> src/by_reference_doc/artists.jl
new file:   src/by_reference_doc/categories.jl
modified:   src/by_reference_doc/episodes.jl
modified:   src/by_reference_doc/player.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/search.jl
modified:   src/by_reference_doc/shows.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/by_reference_doc/users.jl
modified:   src/export_structure.jl
new file:   src/lookup_containers/all_scopes_vector.jl
new file:   src/lookup_containers/response_codes_dic.jl
new file:   src/request.jl
modified:   src/types.jl
deleted:    src/util/request.jl
modified:   src/util/utilities.jl
modified:   test/Project.toml
new file:   test/function_default_arguments_dic.jl
modified:   test/generalize_calls.jl
modified:   test/interactive_default_calls.jl
modified:   test/runtests.jl
modified:   test/test_browse.jl
modified:   test/test_int_format_strings.jl
modified:   test/test_users.jl

hustf committed:
Backup of WIP before picking up.
Update to HTTP > 1.0.
Fix the authorization, scopes and browser launch.
Embrace opt. keywords. When value "", don't send.
              E.g. 'market' is seldom necessary.
Drop arguments like 'market'. Use opt. keywords.
Temp:Show comm. in grey. Plan:Global logswitch.
Postponed until logswitch: Update inline doc.
Move typical call string construction to utils.
Folders by_reference_doc and by_console_doc.
Functions can specify several scopes.
More ducktyping. Both strings and e.g. SpId.
Interpolation: 'url' vs 'body' differ.
TerminalMenus now in source, popul. automatic.

renamed:    readme.md -> README.md. Extend example.
new file:   example/audio_analysis_of_current_track.jl
modified:   src/Spotify.jl     Refer REPL.TerminalMenus, export modules
modified:   src/authorization/access_local_credentials.jl path, message
modified:   src/authorization/client_credentials.jl     ConnectionError
modified:   src/authorization/implicit_grant_flow.jl  No throw, message
modified:   src/by_console_doc/browse.jl        Add recommendations_get
modified:   src/by_console_doc/follow.jl     follow_artists->follow_get
modified:   src/by_console_doc/library.jl       follow_artists_users ->
                                                       follow_get_users
                                    temp disable unfollow_artists_users
modified:   src/by_reference_doc/albums.jl                      General
modified:   src/by_reference_doc/artists.jl                     General
modified:   src/by_reference_doc/episodes.jl                    General
modified:   src/by_reference_doc/genres.jl                      General
modified:   src/by_reference_doc/markets.jl                     General
modified:   src/by_reference_doc/player.jl   Add player_resume_playback
modified:   src/by_reference_doc/playlists.jl                   General
modified:   src/by_reference_doc/search.jl                      General
modified:   src/by_reference_doc/shows.jl             Add working scope
modified:   src/by_reference_doc/tracks.jl   tracks_get_recommendations
modified:   src/by_reference_doc/users.jl                       General
modified:   src/export_structure.jl                export new functions
renamed:    src/lookup_containers/all_scopes_vector.jl -> src/lookup/all_scopes_vector.jl
renamed:    test/parameter_default_dic.jl -> src/lookup/paramname_default_dic.jl
renamed:    src/lookup_containers/response_codes_dic.jl -> src/lookup/response_codes_dic.jl
modified:   src/request.jl   Additional scope. Modify PUT, DELETE, POST
modified:   src/types.jl                                   Check length
new file:   src/util/lenient_conversion_to_string.jl url vs body interp
modified:   src/util/utilities.jl                    Menu auto populate
modified:   test/function_default_arguments_dic.jl       Obsolete, dead
modified:   test/generalize_calls.jl                        Path change
modified:   test/parameter_tuples_vec.jl                 Obsolete, dead
modified:   test/test_browse.jl                                 General
modified:   test/verb_vec.jl                             Obsolete, dead

hustf committed:
modified: src/Spotify.jl TODO, LOG. Less defa scope
modified: src/authorization/implicit_grant_flow.jl

Ask for the existing scopes as well.

modified:  src/by_reference_doc/player.jl add skip_to...
modified:  src/export_structure.jl add skip_to...
modified:  src/request.jl

Respect logstate, optional override logstate

modified:  src/util/lenient_conversion_to_string.jl

Add inline doc

modified: src/util/utilities.jl

Return result if one call only was made.

hustf committed:
modified:   example/mini_player.jl Use LOGSTATE, use Replmode
modified:   src/Spotify.jl   use LOGSTATE
modified:   src/by_reference_doc/playlists.jl update scope
modified:   src/request.jl  use LOGSTATE

hustf committed:
modified:   example/mini_player.jl Implement delete track
modified:   src/by_reference_doc/playlists.jl

            Add playlist_remove_playlist_item

modified:   src/export_structure.jl

            Add playlist_remove_playlist_item

modified:   src/request.jl  Enable DELETE method
modified:   src/util/lenient_conversion_to_string.jl

            Add bodystring(s::T) where T<:Number

hustf committed:
modified:   example/mini_player.jl

    Move replmode out, internal JSON3 reference,
	improve messages, add pause/unpause.

new file:   example/mini_player_replmode.jl    Add 'p'

modified:   src/authorization/access_local_credentials.jl

    Skip scopes if expired credentials.

modified:   src/authorization/implicit_grant_flow.jl Improve message
modified:   src/by_reference_doc/player.jl Add player_pause
modified:   src/export_structure.jl Add player_pause
modified:   src/request.jl Return early if authorization missing.
modified:   test/runtests.jl refresh_spotify_credentials -> authorize

hustf committed:
modified:   example/mini_player.jl Add current_playlist
modified:   example/mini_player_replmode.jl Add current_playlist
modified:   src/Spotify.jl Modify default scope, remove refresh_..
modified:   src/authorization/client_credentials.jl

    refresh_spotify_credentials -> authorize
	Return Object, not nothing

modified:   src/authorization/implicit_grant_flow.jl

    Drop redundant info

hustf committed:
modified:   example/mini_player.jl Add current_audio_features

modified:   example/mini_player_replmode.jl

    Add current_audio_features.
	Fix bug affecting other repl modes.

modified:   src/Spotify.jl

    Avoid duplicate authorization process.

modified:   src/authorization/client_credentials.jl Fix whitespace
modified:   src/by_reference_doc/tracks.jl Document audio features.
modified:   src/util/utilities.jl Drop message when no calls made.

hustf committed:
modified:   Project.toml Drop TerminalMenus, delete compats.
modified:   example/audio_analysis_of_current_track.jl

    Add a note

modified:   example/csv_of_liked_tracks.jl  Cleanup, add note
modified:   src/Spotify.jl                       export JSON3
modified:   src/by_console_doc/library.jl   Default market=""
modified:   src/by_reference_doc/albums.jl  Default market=""
modified:   src/by_reference_doc/episodes.jl Default market
modified:   src/by_reference_doc/playlists.jl Default market
modified:   src/by_reference_doc/shows.jl   Default market=""
modified:   src/by_reference_doc/tracks.jl  Default market=""
modified:   src/util/utilities.jl Excl JSON3 from module menu.
modified:   test/calls.jl                   Default market=""

hustf committed:
modified:   Project.toml Drop Parameters, URIs, StructTypes
modified:   example/audio_analysis_of_current_track.jl Spaces
modified:   example/csv_of_liked_tracks.jl Spaces, explanation
modified:   example/list_my_tracks_beat_per_minute.jl Spaces, explanation
modified:   example/mini_player.jl Spaces
modified:   example/mini_player_replmode.jl Spaces
modified:   example/script_create_playlists_from_referred_tracks_by_bpm.jl

    Formatting

modified:   src/Spotify.jl

    Drop Parameters, URIs, StructTypes
	Update TODO, document LOGSTATE

modified:   src/authorization/access_local_credentials.jl Spaces
modified:   src/authorization/client_credentials.jl Improve 'info'
modified:   src/authorization/implicit_grant_flow.jl

    Refer functions through HTTP. Spaces.

modified:   src/by_console_doc/browse.jl Ref. Spotify.jl:TODO.
modified:   src/by_console_doc/follow.jl Spaces
modified:   src/by_console_doc/library.jl Spaces
modified:   src/by_console_doc/personalization.jl Spaces
modified:   src/by_reference_doc/albums.jl  Spaces
modified:   src/by_reference_doc/artists.jl Ref. Spotify.jl:TODO.
modified:   src/by_reference_doc/episodes.jl Spaces
modified:   src/by_reference_doc/player.jl Ref. Spotify.jl:TODO.
modified:   src/by_reference_doc/playlists.jl Ref. Spotify.jl:TODO.
modified:   src/by_reference_doc/search.jl Spaces
modified:   src/by_reference_doc/shows.jl Spaces
modified:   src/by_reference_doc/tracks.jl  Spaces
modified:   src/by_reference_doc/users.jl Internal func name change.
modified:   src/lookup/paramname_default_dic.jl "no"-> "", "en"->""
modified:   src/request.jl Spaces, docs.
modified:   src/util/lenient_conversion_to_string.jl

    Change name delimit -> build_query_string
	Drop debug code. Spaces.

modified:   src/util/utilities.jl New func assert_locale. Spaces.
modified:   test/calls.jl  Spaces
modified:   test/code_coverage_summary.jl Spaces
deleted:    test/function_default_arguments_dic.jl
deleted:    test/generalize_calls.jl Not used already
deleted:    test/interactive_default_calls.jl Not used already
deleted:    test/parameter_tuples_vec.jl Not used already
modified:   test/test_browse.jl Spaces
deleted:    test/verb_vec.jl Not used already

hustf committed:
Reflect the spec structure more directly.

Add aliases in Library for new functions:
 library_remove_albums +> album_remove_from_library
 library_remove_shows +> shows_remove_from_library
 library_remove_shows +> tracks_remove_from_library
 library_save_album   +> album_save_library
 library_save_shows   +> album_save_library
 library_save_tracks   +> tracks_save_library

Move from Follow to Users:
 follow_check          -> users_check_current_follows
 follow_check_playlist -> users_check_follows_playlist
 follow_get            -> users_get_follows
 follow_get_users      -> users_follow_artists_users
 follow_playlist       -> users_follow_playlist
 unfollow_artists_users-> users_unfollow_artists_users
 unfollow_playlist     -> users_unfollow_playlist

Move from Personalization to Users:
    top_artists         -> users_get_current_user_top_items
    top_tracks          -> users_get_current_user_top_items

modified:   README.md
modified:   docs/src/endpoints.md
modified:   docs/src/index.md
modified:   src/Spotify.jl
modified:   src/by_console_doc/README.md
deleted:    src/by_console_doc/browse.jl
deleted:    src/by_console_doc/follow.jl
modified:   src/by_console_doc/library.jl
deleted:    src/by_console_doc/personalization.jl
modified:   src/by_reference_doc/albums.jl
modified:   src/by_reference_doc/categories.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/shows.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/by_reference_doc/users.jl
modified:   src/export_structure.jl
modified:   test/calls.jl
renamed:    test/test_browse.jl -> test/test_categories.jl
modified:   test/test_tracks.jl
modified:   test/test_users.jl

hustf committed:
modified:   Project.toml Add Logging
modified:   example/audio_analysis_of_current_track.jl

   Check isempty(json)

modified:   example/mini_player.jl    Check isempty(json)
modified:   example/script_create_playlists_from_referred_tracks_by_bpm.jl

  Check isempty(json)

modified:   src/Spotify.jl Use logging
modified:   src/authorization/client_credentials.jl maxlog=1
modified:   src/authorization/implicit_grant_flow.jl

  Increase timeout
  Silence HTTP obnoxious server message
  Drop Windows info about which browser

modified:   src/by_reference_doc/albums.jl  Optional kwds
modified:   src/by_reference_doc/episodes.jl Optional kwds
modified:   src/by_reference_doc/search.jl Optional kwds
modified:   src/by_reference_doc/shows.jl Optional kwds
modified:   src/by_reference_doc/users.jl

    Add scope to users_get_follows
	Optional kwds still missing

modified:   test/calls.jl  Comment it's obsolete
modified:   test/runtests.jl Drop browse.jl, drop authorization
modified:   test/test_albums.jl Separate using
modified:   test/test_artist.jl Separate using, fix test
modified:   test/test_categories.jl Separate using
modified:   test/test_episodes.jl Separate using
modified:   test/test_genres.jl Separate using
modified:   test/test_markets.jl Separate using
modified:   test/test_player.jl Separate using
modified:   test/test_playlists.jl Separate using, fix test
modified:   test/test_search.jl Separate using, fix test
modified:   test/test_shows.jl Separate using, fix test
modified:   test/test_tracks.jl Separate using
modified:   test/test_users.jl Separate using, fix test

hustf committed:
Add user country in .ini file. For default arguments.
modified:   example/audio_analysis_of_current_track.jl Spacing
modified:   example/mini_player_replmode.jl            Spacing
modified:   src/Spotify.jl                                TODO
modified:   src/authorization/client_credentials.jl get_user_country
modified:   src/by_reference_doc/albums.jl               Standardize
modified:   src/by_reference_doc/artists.jl            Standardize
modified:   src/by_reference_doc/categories.jl          Standardize
modified:   src/by_reference_doc/episodes.jl             Standardize
modified:   src/by_reference_doc/genres.jl                Spacing
modified:   src/by_reference_doc/markets.jl               Spacing
modified:   src/by_reference_doc/player.jl                Spacing
modified:   src/by_reference_doc/playlists.jl             Spacing
modified:   src/by_reference_doc/search.jl               Standardize
modified:   src/by_reference_doc/shows.jl              Standardize
modified:   src/by_reference_doc/tracks.jl             Standardize
modified:   src/by_reference_doc/users.jl              Standardize
modified:   src/lookup/paramname_default_dic.jl   get_user_country
modified:   src/request.jl                        Fix for "DELETE"
modified:   src/util/lenient_conversion_to_string.jl  WIP bodystring
modified:   src/util/utilities.jl                     WIP debug
modified:   test/calls.jl                          Func name changes
modified:   test/test_albums.jl   Test no market argument
modified:   test/test_artist.jl  Test no market argument
modified:   test/test_categories.jl  Test no market argument
modified:   test/test_episodes.jl  Test no market argument
modified:   test/test_genres.jl  Func ref change
modified:   test/test_markets.jl  Func ref change
modified:   test/test_player.jl  Func ref change
modified:   test/test_playlists.jl  Test no market argument
modified:   test/test_search.jl  Func ref change
modified:   test/test_shows.jl  Test no market argument
modified:   test/test_tracks.jl  Test no market argument
modified:   test/test_users.jl  Func ref change

hustf committed:
Revise type system, use with duck typing in all calls. See TODO
list in Spotify.jl. All tests passing.

modified:   Project.toml
modified:   example/script_create_playlists_from_referred_tracks_by_bpm.jl
modified:   src/Spotify.jl
modified:   src/authorization/implicit_grant_flow.jl
modified:   src/by_reference_doc/albums.jl
modified:   src/by_reference_doc/artists.jl
modified:   src/by_reference_doc/categories.jl
modified:   src/by_reference_doc/episodes.jl
modified:   src/by_reference_doc/player.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/search.jl
modified:   src/by_reference_doc/shows.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/by_reference_doc/users.jl
modified:   src/lookup/paramname_default_dic.jl
modified:   src/request.jl
modified:   src/types.jl
modified:   src/util/lenient_conversion_to_string.jl
modified:   src/util/utilities.jl
deleted:    test/calls.jl
modified:   test/runtests.jl
modified:   test/test_albums.jl
modified:   test/test_artist.jl
modified:   test/test_categories.jl
modified:   test/test_episodes.jl
modified:   test/test_int_format_strings.jl
modified:   test/test_player.jl
modified:   test/test_playlists.jl
modified:   test/test_search.jl
modified:   test/test_shows.jl
modified:   test/test_tracks.jl
modified:   test/test_users.jl

hustf committed:
Move API ref inside doc., clean whitespace
modified:   README.md        Update
modified:   example/mini_player.jl
modified:   src/Spotify.jl
modified:   src/authorization/access_local_credentials.jl
modified:   src/authorization/implicit_grant_flow.jl
modified:   src/by_console_doc/library.jl
modified:   src/by_reference_doc/albums.jl
modified:   src/by_reference_doc/artists.jl
modified:   src/by_reference_doc/categories.jl
modified:   src/by_reference_doc/episodes.jl
modified:   src/by_reference_doc/genres.jl
modified:   src/by_reference_doc/markets.jl
modified:   src/by_reference_doc/player.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/search.jl
modified:   src/by_reference_doc/shows.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/by_reference_doc/users.jl
modified:   src/request.jl
modified:   src/util/lenient_conversion_to_string.jl
modified:   src/util/utilities.jl

hustf committed:
Fully functional and tested. See Spotify.jl TODO.
modified:   README.md
renamed:    example/mini_player_replmode.jl -> example/_mini_player_replmode.jl
deleted:    example/audio_analysis_of_current_track.jl
modified:   example/list_my_tracks_beat_per_minute.jl
modified:   example/mini_player.jl
new file:   example/playlist_and_library_utilties.jl
new file:   example/remove_duplicates_in_same_playlist.jl
modified:   example/script_create_playlists_from_referred_tracks_by_bpm.jl
modified:   example/script_list_liked_tracks.jl
modified:   src/Spotify.jl
modified:   src/authorization/access_local_credentials.jl
modified:   src/by_reference_doc/albums.jl
modified:   src/by_reference_doc/player.jl
modified:   src/by_reference_doc/playlists.jl
modified:   src/by_reference_doc/tracks.jl
modified:   src/export_structure.jl
modified:   src/lookup/paramname_default_dic.jl
modified:   src/types.jl
modified:   src/util/lenient_conversion_to_string.jl
new file:   src/util/parse_copied_link.jl
modified:   src/util/utilities.jl
modified:   test/runtests.jl
modified:   test/test_int_format_strings.jl
modified:   test/test_player.jl
modified:   test/test_playlists.jl
modified:   test/test_tracks.jl

hustf committed:
modified:   README.md      Typos
modified:   src/Spotify.jl Drop TODO

hustf committed:
Merge branch 'issue53' of https://github.com/hustf/Spotify.jl into issue53

hustf committed:
Adresses issue 53. Tests working on Mac and Windows.

User permissions

`using Spotify` will now ask user for limited privileges through
clicking in a pop-up browser.

-["user-read-private", "user-modify-playback-state",
    "user-read-playback-state", "playlist-modify-private",
    "playlist-read-private"]

If user makes calls which need more, the list is extended and
new user permission is applied for. Grants expire after one
hour. User is warned to call `authorize()`.

Log control, general requests.

- Introduce detail log control, hide non-error details as default.
- HTTP.jl blabber is silenced, which requires adding 'Logging'
from Base.
- For error codes returned, show both the general error
codes and their meaning. Also show detail feedback
which is included in some error message bodies. When errors
are logged (in red), the request is also shown (in red).
- Empty arguments (e.g. 'country=""&offset=0') are not sent. See 'Argument types' below.

`Market`, `locale` and `country` arguments

Where such arguments are necessary, default values are taken from the .ini file, which has a new field for this. For other functions, the default behavior is not to send these arguments.

Menus and maintainability

A rudimentary menu existed, but required maintenance. Menu
now in src, using introspection to build menu.Menu's
argument values are now based on names, not types. Output
is formatted for making inline docs examples.

Argument types

Use duck-typing in all function wrappers, but internal conversion
to SpArtistId, etc. Sp___Id types have one representation
in url strings, and one including type in body strings. Export function `parse_copied_link', which finds type.

Module structure

Move wrapper functions to mirror the 'Endpoints' hierarcy.
A few are only found in the 'Console' hierarchy. These get
alias functions.

Inline docs

Consistent formatting. Will hopefully work with docbuild, but
the 'docs' folder is hardly touched and might require work.

Examples folder, add files

- `playlist_and_library_utilties.jl` for recurring functions.
- `mini_player.jl` ( ':' to enter REPL mode).
- `remove_duplicates_in_same_playlists.jl`

hustf commented:
Fixes issue #53.

hustf commented:
Trying to install on a new machine, I found an issue with writing the default .ini file - it seems to be in JSON format. Also, for those who have existing .ini-files, warnings ought to be added regaring adding 'user' and 'country' files. Will update this PR shortly.

hustf committed:
modified:   src/authorization/client_credentials.jl

Warn if .ini fields missing.
Avoid writing ini file as JSON.

modified:   src/authorization/implicit_grant_flow.jl
   Include 32-bit path for old windows versions.

hustf commented:
Documentation build will still fail. It runs on Julia 1.6. 'Iterators.Pairs' was evidently moved to 'Base.Pairs'. I suggest using a newer version.

kwehmeyer commented on Project.toml:
```diff
@@ -1,27 +1,18 @@
 uuid = "369d3230-709b-4770-b64e-2fec7152d6dc"
-authors = ["Kahlil Wehmeyer","Vikas Negi, PhD", "Owen Dostie"]
-version = "0.1.2"
+authors = ["Kahlil Wehmeyer", "Vikas Negi, PhD", "Owen Dostie"]
```
@hustf feel free to add yourself to this list :)

hustf commented:
Thanks for the approval, Kwehmeyer! I'll be honored to put myself in the list!

Purpose? That's deep. I need to think about that and come back. Also, I'd appreciate your thoughts regarding what to do about the documentation? 

Meanwhile, in the web of commits and regrets comprising this PR, this [summary](https://github.com/kwehmeyer/Spotify.jl/commit/a47210660e523483c4e711e8abfbe3f734fa0d83) got stuck and buried. Here, I pulled it out:


## User permissions

`using Spotify` will now ask user for limited privileges through
clicking in a pop-up browser.

-["user-read-private", "user-modify-playback-state",
    "user-read-playback-state", "playlist-modify-private",
    "playlist-read-private"]

If user makes calls which need more, the list is extended and
new user permission is applied for. Grants expire after one
hour. User is warned to call `authorize()`.


## Log control, general requests.

- Introduce detail log control, hide non-error details as default.
- HTTP.jl blabber is silenced, which requires adding 'Logging'
from Base.
- For error codes returned, show both the general error
codes and their meaning. Also show detail feedback
which is included in some error message bodies. When errors
are logged (in red), the request is also shown (in red).
- Empty arguments (e.g. 'country=""&offset=0') are not sent. See 
  'Argument types' below.


## `Market`, `locale` and `country` arguments

Where such arguments are necessary, default values are taken from the .ini file, 
which has a new field for this. For other functions, the default behavior is not to 
send these arguments.


## Menus and maintainability

A rudimentary menu existed, but required maintenance. Menu
now in src, using introspection to build menu.Menu's
argument values are now based on names, not types. Output
is formatted for making inline docs examples.


## Argument types

Use duck-typing in all function wrappers, but internal conversion
to SpArtistId, etc. Sp___Id types have one representation
in url strings, and one including type in body strings. Export function 
`parse_copied_link', which finds type.


## Module structure

Move wrapper functions to mirror the 'Endpoints' hierarcy.
A few are only found in the 'Console' hierarchy. These get
alias functions.


## Inline docs

Consistent formatting. Will hopefully work with docbuild, but
the 'docs' folder is hardly touched and might require work.

Examples folder, add files

- `playlist_and_library_utilties.jl` for recurring functions.
- `mini_player.jl` ( ':' to enter REPL mode).
- `remove_duplicates_in_same_playlists.jl`

hustf commented:
# My purpose 

with this submission is to prepare for a public transport regularity analysis.
These parts are quite reusable:

- The limited type system, where representation depends on context: JSON, terminal, 
    or URI.
- One 'request' function for all endpoints, including authorization fields.
- Duck typing with internal conversion.

# Package's purpose 

in pulling this would be to improve usability to that stage where any Julia user 
can easily adapt examples to fit in ones own algorithms.

To this purpose, every call outputs the tuple '(Json3Object, Integer)' as before. Examples are 
one level up from the web API interface. This level is where user 
 - parses returned JSON3Object, perhaps to Spotify.jl's types.
 - stores in Dict, LightXML, DataFrame or maybe jute sacks? 

# Problem fixes

 - ❌ means 'has a problem' or 'not working'
- ✓  means 'fixed'
  - ⌅ means 'limits to'
 - => means 'leads to'


| Problems as experienced on Windows 11  |v0.1.2| v0.2 | Ref.                 |
| :---                                   |:---: | :---:| :---                 |
| test_playlists.jl test_player.jl       |  ❌  |   ✓  |                     |
| test_tracks.jl test_browse.jl          |  ❌  |   ✓  |                     |
| test_episodes.jl                       |  ❌  |   ✓  |                     |
| HTTP ⌅ 0.9.17 => ⌅Compat ⌅TerminalMenus|  ❌  |   ✓  | Project.toml        |
| Warnings: Could not export etc..       |  ❌  |   ✓  | export_structure.jl |
| Browser launch not working             |  ❌  |   ✓  |implicit_grant_flow.jl|
| Default is 'Client credentials' only   |  ❌  |   ✓  | 1) below            |
| Feedback from errored calls partly lost|  ❌  |   ✓  |util/request.jl -> src|
| Some function names may confuse        |  ❌  |   ✓  | 2)                  |
| Web docs have several hierarchies      |  ❌  |   ❌ | 3)                  |
| 'csv_of_saved_tracks.jl' (master only) |  ❌   |  ✓  | 4)                  |
| Terminal messages not targeted to user |  ❌   |  ✓  |?LOGSTATE, request.jl|
| Request body feature for web API calls |  ❌   |  ✓  | 5)                  |
| Some country / market args hardcoded   |  ❌   |  ✓  | .ini field, also drop args|


1)  'Client credentials' require no user grants and is quite limited. Default 
   rate limits are severe, but less so with a user grant. It will be hard
   to do useful analysis without implicit grants. Hence, we now immediately 
   `apply_and_wait_for_implicit_grant(;scopes =..)`

2) Function names now describe what they do in calling context:
   -`implicit_grant_flow/request_handler` -> `receive_grant_as_request`
   - `start_a_listening_server`  -> `launch_async_single_grant_receiving_server`


3) Considering: 
    - 'reference/endpoints' is the complete hierarchy.
    - The 'Console' tree contains the necessary grant scopes.
    
  File folders
    - 'by_reference_doc' folder now accurately reflects the former.
    - 'by_console_doc' contain useful alias functions. Deleted file 'follow.jl',
       whose functions now rests in 'users.jl'.

4) This (unreleased) example required running in a separate environment with 
   DataFrames and CSV. But HTTP ⌅ 0.9.17 also limited these packages to ancient
   versions.

   Renamed file to 'csv_of_liked_tracks.jl', added explaining comments, pruned bitrot.

5) Some calls can convey arguments through urlstring as an alternative to a body.
    Usually, urlstring has more severe limitations. Wrappers which now sends arguments as a body:
    - `users_unfollow_artists_users`
    - `users_follow_playlist`
    - `playlist_create_playlist`
    - `playlist_remove_playlist_item`
    - `playlist_add_tracks_to_playlist`
    - `playlist_change_details`
    - `player_resume_playback`

hustf committed:
Minimal changes for a local docbuild.

modified:   .github/workflows/builddoc.yml Remove J v1.6 requirement
modified:   Project.toml    Add HustF
modified:   docs/Project.toml Add HTTP, IniFile, JSON3/StructTypes

hustf commented:
The latest commit added me to the authors list as recommended, and made minimal changes for a successful docbuild.
I believe this is ready to be merged, but I don't have write access to do so.
